File: Algorithm/Java/src/main/java/com/yugutou/charpter19_dp/level2/LengthOfLIS.java
package com.yugutou.charpter19_dp.level2;

/**
 * LeetCode300.
 * 给你一个整数数组 nums ，找到其中最长严格递增子序列的长度。
 */
public class LengthOfLIS {
    public static void main(String[] args) {
        int[] nums = {10, 9, 2, 5, 3, 7, 101, 18};
        System.out.println(lengthOfLIS(nums));
    }

    public static int lengthOfLIS(int[] A) {
        int n = A.length;
        if (n == 0) {
            return 0;
        }
        int[] f = new int[n];
        int i, j, res = 0;
        for (j = 0; j < n; j++) {
            f[j] = 1;
            for (i = 0; i < j; i++) {
                if (A[i] < A[j] && f[i] + 1 > f[j]) {
                    f[j] = f[i] + 1;
                }
            }
            res = Math.max(res, f[j]);
        }
        return res;
    }
}

File: Algorithm/Leetcode/Code_Materials/c_or_cpp/charpter19_dp/level2/length_of_LIS.cpp
#include <iostream>
#include <vector>

using namespace std;

int lengthOfLIS(vector<int>& A) {
    int n = A.size();
    if (n == 0) {
        return 0;
    }
    int f[n];
    int i, j, res = 0;
    for (j = 0; j < n; j++) {
        f[j] = 1;
        for (i = 0; i < j; i++) {
            if (A[i] < A[j] && f[i] + 1 > f[j]) {
                f[j] = f[i] + 1;
            }
        }
        res = max(res, f[j]);
    }
    return res + 1;
}

int main() {
    vector<int> A = {1, 3, 5, 7, 9};
    cout << lengthOfLIS(A) << endl;   
    return 0;
}